File: Wikipedia Maçkolik-TFF Match Scraper/NewFolder1/StadiumPlace.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace Wikipedia_Maçkolik_TFF_Match_Scraper.NewFolder1
{
    public class StadiumPlace
    {

        public string QIDCity(int id)
        {
            HttpClient httpClient = new HttpClient();
            httpClient.DefaultRequestHeaders.Add("User-Agent", "PostmanRuntime/7.34.0");

            File.WriteAllText("a.txt", $"https://query.wikidata.org/sparql?query=SELECT%20%3Fitem%0AWHERE%20%7B%0A%20%20%3Fitem%20wdt%3AP7402%20\"{id}\"%20.%0A%7D&format=json");
            string myJsonResponse = httpClient.GetStringAsync($"https://query.wikidata.org/sparql?query=SELECT%20%3Fitem%0AWHERE%20%7B%0A%20%20%3Fitem%20wdt%3AP7402%20\"{id}\"%20.%0A%7D&format=json").Result;
            JObject jObject = JObject.Parse(myJsonResponse);
            var qid = jObject["results"]["bindings"][0]["item"]["value"].ToString().Replace("http://www.wikidata.org/entity/", "");

            string myJsonResponse2 = httpClient.GetStringAsync($"https://www.wikidata.org/w/api.php?action=wbgetentities&ids=" + jObject["results"]["bindings"][0]["item"]["value"].ToString().Replace("http://www.wikidata.org/entity/", "") + "&format=json").Result;
            JObject jObject2 = JObject.Parse(myJsonResponse2);

            try
            {

                var a = jObject2["entities"][qid]["claims"]?["P131"][0]["mainsnak"]?["datavalue"]?["value"]?["id"].ToString();
                if (a.Contains("("))
                    return a + "|" + a.Substring(0, a.IndexOf("(") - 1);
                else
                    return a;

            }
            catch (Exception)
            {
                Console.WriteLine("Stadın şehri wikidatada yok -- {0}", id);

            }
            return null;

        }

        public string QID(int id)
        {
            HttpClient httpClient = new HttpClient();
            httpClient.DefaultRequestHeaders.Add("User-Agent", "PostmanRuntime/7.34.0");

            //File.WriteAllText("a.txt", $"https://query.wikidata.org/sparql?query=SELECT%20%3Fitem%0AWHERE%20%7B%0A%20%20%3Fitem%20wdt%3AP7402%20\"{id}\"%20.%0A%7D&format=json");
            //string myJsonResponse = httpClient.GetStringAsync($"https://query.wikidata.org/sparql?query=SELECT%20%3Fitem%0AWHERE%20%7B%0A%20%20%3Fitem%20wdt%3AP131%20\"{id}\"%20.%0A%7D&format=json").Result;
            ////Match myDeserializedClass = JsonConvert.DeserializeObject<Match>(myJsonResponse);
            //JObject jObject = JObject.Parse(myJsonResponse);
            //var qid = jObject["results"]["bindings"][0]["item"]["value"].ToString().Replace("http://www.wikidata.org/entity/", "");
            try
            {

            var qid = QIDCity(id);
            string myJsonResponse2 = httpClient.GetStringAsync($"https://www.wikidata.org/w/api.php?action=wbgetentities&ids=" + qid + "&format=json").Result;
            JObject jObject2 = JObject.Parse(myJsonResponse2);
            try
            {
                if (jObject2["entities"][qid]["sitelinks"]?["trwiki"]?["title"].ToString() != null)
                {
                    var a = jObject2["entities"][qid]["sitelinks"]?["trwiki"]?["title"].ToString();
                    if (a.Contains("("))
                        return a + "|" + a.Substring(0, a.IndexOf("(") - 1);
                    else
                        return a;
                }
                else
                {
                    var a = jObject2["entities"][qid]["sitelinks"]?["enwiki"]?["title"].ToString();
                    if (a.Contains("("))
                        return a + "|" + a.Substring(0, a.IndexOf("(") - 1);
                    else
                        return a;
                }
            }
            catch (Exception)
            {
                if (jObject2["entities"][qid]["labels"]["tr"]["value"].ToString() != null)
                {
                    var a = jObject2["entities"][qid]["labels"]["tr"]["value"].ToString();
                    if (a.Contains("("))
                        return a + "|" + a.Substring(0, a.IndexOf("(") - 1);
                    else
                        return a;
                }
            }

            }
            catch (Exception)
            {


            }
            return id.ToString();
        }
    }
}

File: Wikipedia Maçkolik-TFF Match Scraper/docs/js/wikidata/StadiumPlace.js
/**
 * StadiumPlace - Resolves stadium location (city) from Wikidata using TFF stadium ID
 */
class StadiumPlace {
    constructor() {
        this.cache = new Map();
    }

    /**
     * Get city QID from stadium
     * @param {number} id - TFF stadium ID
     * @returns {Promise<string|null>} - City QID
     */
    async getCityQID(id) {
        try {
            // Query Wikidata for entity with TFF stadium ID (P7402)
            const sparqlQuery = `SELECT ?item WHERE { ?item wdt:P7402 "${id}" . }`;
            const sparqlUrl = `https://query.wikidata.org/sparql?query=${encodeURIComponent(sparqlQuery)}&format=json`;

            const response = await fetch(sparqlUrl, {
                headers: {
                    'User-Agent': 'WikipediaMatchScraper/1.0'
                }
            });

            if (!response.ok) throw new Error('SPARQL query failed');

            const data = await response.json();

            if (!data.results.bindings.length) {
                return null;
            }

            const entityUrl = data.results.bindings[0].item.value;
            const qid = entityUrl.replace('http://www.wikidata.org/entity/', '');

            // Get entity details to find P131 (located in administrative territorial entity)
            const entityResponse = await fetch(
                `https://www.wikidata.org/w/api.php?action=wbgetentities&ids=${qid}&format=json&origin=*`
            );

            if (!entityResponse.ok) throw new Error('Entity fetch failed');

            const entityData = await entityResponse.json();
            const entity = entityData.entities[qid];

            // Get P131 claim (located in)
            if (entity.claims?.P131?.[0]?.mainsnak?.datavalue?.value?.id) {
                return entity.claims.P131[0].mainsnak.datavalue.value.id;
            }

            return null;

        } catch (error) {
            console.error(`Error fetching city for stadium ID ${id}:`, error);
            return null;
        }
    }

    /**
     * Get stadium location name from Wikidata by TFF stadium ID
     * @param {number} id - TFF stadium ID
     * @returns {Promise<string>} - Location name for Wikipedia
     */
    async getStadiumPlace(id) {
        // Check cache first
        if (this.cache.has(id)) {
            return this.cache.get(id);
        }

        try {
            const cityQID = await this.getCityQID(id);

            if (!cityQID) {
                this.cache.set(id, id.toString());
                return id.toString();
            }

            // Get city entity details
            const entityResponse = await fetch(
                `https://www.wikidata.org/w/api.php?action=wbgetentities&ids=${cityQID}&format=json&origin=*`
            );

            if (!entityResponse.ok) throw new Error('Entity fetch failed');

            const entityData = await entityResponse.json();
            const entity = entityData.entities[cityQID];

            // Try to get Turkish Wikipedia article title first, then English, then label
            let name = null;

            if (entity.sitelinks?.trwiki?.title) {
                name = entity.sitelinks.trwiki.title;
            } else if (entity.sitelinks?.enwiki?.title) {
                name = entity.sitelinks.enwiki.title;
            } else if (entity.labels?.tr?.value) {
                name = entity.labels.tr.value;
            } else {
                name = id.toString();
            }

            // Format name with disambiguation if needed
            const formattedName = this.formatName(name);
            this.cache.set(id, formattedName);
            return formattedName;

        } catch (error) {
            console.error(`Error fetching stadium place for ID ${id}:`, error);
            this.cache.set(id, id.toString());
            return id.toString();
        }
    }

    /**
     * Format name with disambiguation pipe if contains parentheses
     */
    formatName(name) {
        if (name.includes('(')) {
            const baseName = name.substring(0, name.indexOf('(') - 1);
            return `${name}|${baseName}`;
        }
        return name;
    }

    /**
     * Clear cache
     */
    clearCache() {
        this.cache.clear();
    }
}

// Export for use in other modules
if (typeof window !== 'undefined') {
    window.StadiumPlace = StadiumPlace;
}
